File: android/src/main/java/com/unact/yandexmapkit/full/YandexDrivingSession.java
package com.unact.yandexmapkit.full;

import androidx.annotation.NonNull;

import com.yandex.mapkit.RequestPoint;
import com.yandex.mapkit.directions.driving.DrivingRoute;
import com.yandex.mapkit.directions.driving.DrivingRouter;
import com.yandex.mapkit.directions.driving.DrivingSession;
import com.yandex.mapkit.directions.driving.VehicleOptions;
import com.yandex.mapkit.directions.driving.Weight;
import com.yandex.runtime.Error;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.Result;

public class YandexDrivingSession implements MethodChannel.MethodCallHandler {
  private final int id;
  private DrivingSession session;
  private final MethodChannel methodChannel;
  private final DrivingRouter drivingRouter;
  @SuppressWarnings({"MismatchedQueryAndUpdateOfCollection"})
  private static final Map<Integer, YandexDrivingSession> drivingSessions = new HashMap<>();

  public static void initSession(int id, BinaryMessenger messenger, DrivingRouter drivingRouter) {
    drivingSessions.put(id, new YandexDrivingSession(id, messenger, drivingRouter));
  }

  public YandexDrivingSession(
    int id,
    BinaryMessenger messenger,
    DrivingRouter drivingRouter
  ) {
    this.id = id;
    this.drivingRouter = drivingRouter;

    methodChannel = new MethodChannel(messenger, "yandex_mapkit/yandex_driving_session_" + id);
    methodChannel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "requestRoutes":
        requestRoutes(call, result);
        break;
      case "cancel":
        cancel();
        result.success(null);
        break;
      case "retry":
        retry(result);
        break;
      case "close":
        close();
        result.success(null);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void requestRoutes(final MethodCall call, final Result result) {
    YandexDrivingSession self = this;
    Map<String, Object> params = (Map<String, Object>) call.arguments;
    List<RequestPoint> points = new ArrayList<>();
    for (Map<String, Object> pointParams : (List<Map<String, Object>>) params.get("points")) {
      points.add(UtilsFull.requestPointFromJson(pointParams));
    }

    session = drivingRouter.requestRoutes(
      points,
      UtilsFull.drivingOptionsFromJson((Map<String, Object>) params.get("drivingOptions")),
      new VehicleOptions(),
      new DrivingSession.DrivingRouteListener() {
        @Override
        public void onDrivingRoutes(@NonNull List<DrivingRoute> list) { self.onDrivingRoutes(list, result); }
        @Override
        public void onDrivingRoutesError(@NonNull Error error) { self.onDrivingRoutesError(error, result); }
      }
    );
  }

  public void cancel() {
    session.cancel();
  }

  public void retry(MethodChannel.Result result) {
    YandexDrivingSession self = this;

    session.retry(
      new DrivingSession.DrivingRouteListener() {
        @Override
        public void onDrivingRoutes(@NonNull List<DrivingRoute> list) { self.onDrivingRoutes(list, result); }
        @Override
        public void onDrivingRoutesError(@NonNull Error error) { self.onDrivingRoutesError(error, result); }
      }
    );
  }

  public void close() {
    session.cancel();
    methodChannel.setMethodCallHandler(null);

    drivingSessions.remove(id);
  }

  private void onDrivingRoutes(@NonNull List<DrivingRoute> list, @NonNull Result result) {
    List<Map<String, Object>> resultRoutes = new ArrayList<>();
    for (DrivingRoute route : list) {
      Weight weight = route.getMetadata().getWeight();
      Map<String, Object> resultWeight = new HashMap<>();
      resultWeight.put("time", UtilsFull.localizedValueToJson(weight.getTime()));
      resultWeight.put("timeWithTraffic", UtilsFull.localizedValueToJson(weight.getTimeWithTraffic()));
      resultWeight.put("distance", UtilsFull.localizedValueToJson(weight.getDistance()));
      Map<String, Object> resultMetadata = new HashMap<>();
      resultMetadata.put("weight", resultWeight);

      Map<String, Object> resultRoute = new HashMap<>();
      resultRoute.put("geometry", UtilsFull.polylineToJson(route.getGeometry()));
      resultRoute.put("metadata", resultMetadata);

      resultRoutes.add(resultRoute);
    }

    Map<String, Object> resultMap = new HashMap<>();
    resultMap.put("routes", resultRoutes);

    result.success(resultMap);
  }

  private void onDrivingRoutesError(@NonNull Error error, @NonNull Result result) {
    result.success(UtilsFull.errorToJson(error));
  }
}

File: ios/Classes/full/YandexDrivingSession.swift
import Foundation
import YandexMapsMobile

public class YandexDrivingSession: NSObject {
  private var id: Int
  private var session: YMKDrivingSession!
  private let methodChannel: FlutterMethodChannel!
  private let drivingRouter: YMKDrivingRouter!
  private static var drivingSessions: [Int: YandexDrivingSession] = [:]

  public static func initSession(id: Int, registrar: FlutterPluginRegistrar, drivingRouter: YMKDrivingRouter) {
    drivingSessions[id] = YandexDrivingSession(id: id, registrar: registrar, drivingRouter: drivingRouter)
  }

  public required init(id: Int, registrar: FlutterPluginRegistrar, drivingRouter: YMKDrivingRouter) {
    self.id = id
    self.methodChannel = FlutterMethodChannel(
      name: "yandex_mapkit/yandex_driving_session_\(id)",
      binaryMessenger: registrar.messenger()
    )
    self.drivingRouter = drivingRouter

    super.init()

    weak var weakSelf = self
    self.methodChannel.setMethodCallHandler({ weakSelf?.handle($0, result: $1) })
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "requestRoutes":
      requestRoutes(call, result)
    case "cancel":
      cancel()
      result(nil)
    case "retry":
      retry(result)
    case "close":
      close()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func requestRoutes(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    let params = call.arguments as! [String: Any]
    let requestPoints = (params["points"] as! [[String: Any]]).map {
      (pointParams) -> YMKRequestPoint in UtilsFull.requestPointFromJson(pointParams)
    }

    session = drivingRouter.requestRoutes(
      with: requestPoints,
      drivingOptions: UtilsFull.drivingOptionsFromJson(params["drivingOptions"] as! [String: Any]),
      vehicleOptions: YMKDrivingVehicleOptions(),
      routeHandler: {(drivingResponse: [YMKDrivingRoute]?, error: Error?) -> Void in
        self.handleResponse(drivingResponse: drivingResponse, error: error, result: result)
      }
    )
  }

  public func cancel() {
    session.cancel()
  }

  public func retry(_ result: @escaping FlutterResult) {
    session.retry(routeHandler: {(drivingResponse: [YMKDrivingRoute]?, error: Error?) -> Void in
      self.handleResponse(drivingResponse: drivingResponse, error: error, result: result)
    })
  }

  public func close() {
    session.cancel()

    YandexDrivingSession.drivingSessions.removeValue(forKey: id)
  }

  public func handleResponse(drivingResponse: [YMKDrivingRoute]?, error: Error?, result: @escaping FlutterResult) {
    if let response = drivingResponse {
      onSuccess(response, result)
    } else {
      onError(error!, result)
    }
  }

  private func onSuccess(_ res: [YMKDrivingRoute], _ result: @escaping FlutterResult) {
    let routes = res.map { (route) -> [String: Any] in
      let weight = route.metadata.weight

      return [
        "geometry": UtilsFull.polylineToJson(route.geometry),
        "metadata": [
          "weight": [
            "time": UtilsFull.localizedValueToJson(weight.time),
            "timeWithTraffic": UtilsFull.localizedValueToJson(weight.timeWithTraffic),
            "distance": UtilsFull.localizedValueToJson(weight.distance)
          ]
        ]
      ]
    }

    let arguments: [String: Any] = [
      "routes": routes
    ]

    result(arguments)
  }

  private func onError(_ error: Error, _ result: @escaping FlutterResult) {
    result(UtilsFull.errorToJson(error))
  }
}
